File: py/archbird/_serve_worker.py
"""Isolated Map candidate process for the local visualization host."""

from __future__ import annotations

import base64
import json
from pathlib import Path
import sys

from .native import Project


def main() -> int:
    request = json.loads(sys.stdin.buffer.read())
    root = Path(request["root"]).resolve()
    encoded_config = request.get("config_base64")
    config = base64.b64decode(encoded_config) if encoded_config is not None else None
    if request.get("config_path") is not None:
        config = Path(request["config_path"]).read_bytes()
    if not request.get("no_config") and config is None:
        candidate = root / "archbird.json"
        if candidate.is_file() and not candidate.is_symlink():
            config = candidate.read_bytes()
    options = dict(request.get("project_options") or {})
    project = Project.from_repository(root, config=config, **options)
    map_json = project.map_json()
    document = json.loads(map_json)
    generation = document.get("evidence", {}).get("input_sha256")
    if not isinstance(generation, str) or len(generation) != 64:
        raise RuntimeError("generated Map has no valid evidence.input_sha256")
    header = {
        "files": [
            {
                "bytes": row["bytes"],
                "path": row["path"],
                "sha256": row["sha256"],
            }
            for row in document["files"]
        ],
        "generation": generation,
        "map_bytes": len(map_json),
        "project": document["project"],
    }
    sys.stdout.buffer.write(
        json.dumps(header, ensure_ascii=True, separators=(",", ":"), sort_keys=True).encode()
        + b"\n"
        + map_json
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: js/src/serve-worker.js
"use strict";

const fs = require("node:fs");
const path = require("node:path");
const { parentPort, workerData } = require("node:worker_threads");
const { Project } = require("./index");

function configBytes(root, configured) {
  if (workerData.configJson) return Buffer.from(workerData.configJson);
  if (workerData.noConfig) return null;
  if (configured) return fs.readFileSync(path.resolve(configured));
  const candidate = path.join(root, "archbird.json");
  try {
    const metadata = fs.lstatSync(candidate);
    return metadata.isFile() && !metadata.isSymbolicLink()
      ? fs.readFileSync(candidate)
      : null;
  } catch (error) {
    if (error && error.code === "ENOENT") return null;
    throw error;
  }
}

function main() {
  const root = path.resolve(workerData.root);
  const project = Project.fromRepository(root, {
    config: configBytes(root, workerData.config),
    ...(workerData.options || {}),
    typescript: workerData.typescript,
  });
  try {
    const map = project.mapJson();
    const document = JSON.parse(map.toString("utf8"));
    const generation = document.evidence?.input_sha256;
    if (typeof generation !== "string" || !/^[0-9a-f]{64}$/.test(generation)) {
      throw new Error("generated Map has no valid evidence.input_sha256");
    }
    const bytes = Uint8Array.from(map);
    parentPort.postMessage({
      files: document.files.map((row) => ({
        bytes: row.bytes,
        path: row.path,
        sha256: row.sha256,
      })),
      generation,
      map: bytes.buffer,
      project: document.project,
    }, [bytes.buffer]);
  } finally {
    project.dispose();
  }
}

try {
  main();
} catch (error) {
  const prefixes = [
    `${path.resolve(workerData.root)}${path.sep}`,
    ...(workerData.config
      ? [`${path.dirname(path.resolve(workerData.config))}${path.sep}`]
      : []),
  ];
  let message = String(error && (error.stack || error.message) || error);
  for (const prefix of prefixes) message = message.replaceAll(prefix, "");
  parentPort.postMessage({
    error: message,
  });
}
